src/com/twitter/intellij/pants/bsp/JarMappings.java

```diff
@@ -125,17 +128,27 @@ private String targetAddressFromSanitizedFileName(String jarName) {
 
   private boolean isProjectInternalDependency(VirtualFile jar) {
     Path jarPath = Paths.get(jar.getPath());
-    Path bloopJarsPath = bloopJarsPath(project);
-    return jarPath.startsWith(bloopJarsPath);
+    Optional<Path> bloopJarsPath = bloopJarsPath(project);
+    Optional<Path> sourcesJarPath = sourcesJarPath(project);
+    return jarPath.startsWith(bloopJarsPath.get()) || jarPath.startsWith(sourcesJarPath.get());
   }
 
-  private static Path bloopJarsPath(Project project) {
-    return Paths.get(project.getBasePath(), ".bloop", "bloop-jars");
+  private static Optional<Path> bloopJarsPath(Project project) {
+    return PantsBspData.importsFor(project).stream().findFirst()
+      .map(path -> Paths.get(path.getBspPath().toString(), ".bloop", "bloop-jars"));
   }
 
-  private VirtualFile librariesFile() {
-    Path path = Paths.get(project.getBasePath(), ".pants", "libraries.json");
-    return LocalFileSystem.getInstance().findFileByIoFile(path.toFile());
+  private static Optional<Path> sourcesJarPath(Project project) {
+    return PantsBspData.importsFor(project).stream().findFirst()
```

Comment:
we could have a method to return bloop directory and then this and bloopJarsPath could just resolve single child